content/blog/mixtral-8x7b-healthcare-chatbot/index.md

```diff
@@ -0,0 +1,167 @@
+
+def documentize(article):
+  return Document(content=article.abstract, meta={'title': article.title, 'keywords': article.keywords})
+
+@component
+class PubMedFetcher():
+
+  @component.output_types(articles=List[Document])
+  def run(self, queries: list[str]):
+    cleaned_queries = queries[0].strip().split('\n')
+
+    articles = []
+    try:
+      for query in cleaned_queries:
+        response = pubmed.query(query, max_results = 1)
+        documents = [documentize(article) for article in response]
+        articles.extend(documents)
+    except Exception as e:
+        print(e)
+        print(f"Couldn't fetch articles for queries: {queries}" )
+    results = {'articles': articles}
+    return results
+```
+
+For the model, I went with Mixtral's 8x7b. Mixtral is a unique kind of model that uses 8 *"experts"* and an internal *"routing"* mechanism that routes a token to a specific expert. This also means that during inferencing, not all parameters are used, which allows the model to response remarkably fast.  [This HuggingFace blog post explains MoE](https://huggingface.co/blog/moe) in more detail. 
+
+
+## Generating Keywords for PubMed with Mixtral/LLMs
+
+First, I tried an approach where I passed a plain query to PubMed. e.g. *"What are the most  current treatments for long COVID?"* Unfortunately, that didn't work too well. The articles returned weren't very relevant. Which makes sense, because PubMed isnt optimized for natural language search. It is optimized for keywords, though. And you know what's great at generating keywords? LLMs!
+
+So now our flow is as follows:
+- Our user inputs a question, such as, "
+- We prompt the LLM to turn tha question into keywords
+- Search PubMed and return top_k articles based on those keywords
+- Pass those articles to the LLM and ask them to reference it when formulating an answer.
+
+![](HealthcareChatbotArchitecture.png)
+
+```python
+from haystack import Pipeline
+from haystack.components.builders.prompt_builder import PromptBuilder
+
+keyword_prompt_template = """
+Your task is to convert the follwing question into 3 keywords that can be used to find relevant medical research papers on PubMed.
+Here is an examples:
+question: "What are the latest treatments for major depre###ssive disorder?"
+keywords:
+Antidepressive Agents
+Depressive Disorder, Major
+Treatment-Resistant depression
+---
+question: {{ question }}
+keywords:
+"""
+
+prompt_template = """
+Answer the question truthfully based on the given documents.
+If the documents don't contain an answer, use your existing knowledge base.
+
+q: {{ question }}
+Articles:
+{% for article in articles %}
+  {{article.content}}
+  keywords: {{article.meta['keywords']}}
+  title: {{article.meta['title']}}
+{% endfor %}
+
+"""
+keyword_prompt_builder = PromptBuilder(template=keyword_prompt_template)
+prompt_builder = PromptBuilder(template=prompt_template)
+fetcher = PubMedFetcher()
```

Comment:
I believe the code snippets to initalize the Generators are missing, can you provide codes for them somewhere in the article?